### PR TITLE
Imputation

### LINKED ISSUE #80: imputation based on multiple samples
Our imputation results are based on just one sample from the variational distribution currently:

https://github.com/YosefLab/scVI/blob/704ce19d80e5728542464e488968c136a241eb6f/scvi/metrics/imputation.py#L24

That makes imputation not much of a metric for assessing changes to our model --- I'm surprised it works as well as it does. We should be sure to fix it before relying on imputation scores to guide any modeling decisions.

### PR DESCRIPTION
**Pull request by Edouard360**
resolves #80 and starts #123

From the new implementation, with Cortex [250 epochs], imputation median ~2.3.

imputation-1
 

- imputation:

  - corruption argument: binomial/bernouilli
  - training: before, only a forward pass on corrupted data was done, whereas the model needs to be trained on the corrupted data
  - n_samples: Multiple samples for get_sample_rate (which by the way before was not sampled, but only taking the mean qz_m, ql_m)
  - returning per cell original(but corrupted)/imputed value as output of the imputation
  - In output, print the global median or mean of median (across cells)

- other changes:
  - in modules, the forward pass for sequential layers now handles dim==3, ie. (n_samples, n_batch, n_features).
  - simplification in naming attributes:
    - data_loaders.data_loaders_loop -> data_loaders.loop
    - data_loaders.data_loaders_dict -> data_loaders.dict
    - data_loaders.data_loaders_kwargs -> data_loaders.kwargs
  - had to add a .sequential function to DataLoaderWrapper and a SequentialSubsetSampler, since we'd like to train on shuffled data_loaders, but when we compare corrupted/original values, we iterate jointly on the same indices

### CODE REVIEW

File: scvi/inference/variational_inference.py [hunk 1]
```diff
@@ -67,14 +67,27 @@ def ll(self, name, verbose=False):
...
-    def imputation_errors(self, name, **kwargs):
-        return imputation(self.model, self.data_loaders[name], **kwargs)
+    def imputation(self, name, verbose=False, rate=0.1, n_samples=1, n_epochs=1, corruption="uniform",
+                   title="Imputation"):
+        original_list, imputed_list = imputation(self, name, rate=rate, n_epochs=n_epochs, n_samples=n_samples,
+                                                 corruption=corruption)
+        # Median of medians for all distances
+        median_imputation_score = np.median(np.abs(np.concatenate(original_list) - np.concatenate(imputed_list)))
+
+        # Mean of medians for each cell
```
Comment (Edouard360) [inline scvi/inference/variational_inference.py]: Is this what you had in mind @jeff-regier ?

**Comment by Edouard360** [comment]
@jeff-regier: you mentioned iirc to use the "median per cell" for the genes an then take the mean of that. Whereas the global median gives ~2.3, the mean of medians gives ~4.3. I still output it in `infer.imputation` but I can correct if you meant something else..

File: scvi/metrics/imputation.py [hunk 2]
```diff
@@ -1,18 +1,118 @@
...
+        infer.data_loaders['corrupted_%s' % key] = DataLoaderWrapper(gene_dataset, use_cuda=infer.use_cuda, **kwargs)
+
+    infer.train(n_epochs=n_epochs)
+    infer.data_loaders.loop = original_data_loaders_loop
+
+    original_list = []
+    imputed_list = []
+    batch_size = infer.data_loaders.kwargs["batch_size"] // n_samples
+    for tensors, corrupted_tensors in \
+        zip(infer.data_loaders[name].sequential(batch_size=batch_size),
```
Comment (Edouard360) [inline scvi/metrics/imputation.py]: This is where I used the `.sequential` method and the SequentialSubsetSampler needed to iterate simultaneously on the same indices for both the uncorrupted and corrupted data.

File: scvi/metrics/imputation.py [hunk 3]
```diff
@@ -1,18 +1,118 @@
...
+    original_data_loaders_loop = infer.data_loaders.loop
+    infer.data_loaders.loop = ['corrupted_%s' % s for s in infer.data_loaders.loop]
+    original_keys = list(infer.data_loaders.dict.keys())
+    for key in original_keys:
+        kwargs = copy.copy(infer.data_loaders.kwargs)
+        kwargs['collate_fn'] = gene_dataset.collate_fn
+        kwargs['sampler'] = copy.copy(infer.data_loaders[key].sampler)
+        infer.data_loaders['corrupted_%s' % key] = DataLoaderWrapper(gene_dataset, use_cuda=infer.use_cuda, **kwargs)
+
+    infer.train(n_epochs=n_epochs)
```
Comment (Edouard360) [inline scvi/metrics/imputation.py]: Before there was no training

File: scvi/metrics/imputation.py [hunk 4]
```diff
@@ -1,18 +1,118 @@
...
 import numpy as np
 import torch
+from scipy.stats import kde
+
+from scvi.dataset import GeneExpressionDataset
+from scvi.dataset.data_loaders import DataLoaderWrapper
+
+
+def imputation(infer, name, rate=0.1, n_samples=1, n_epochs=1, corruption="uniform"):
+    corrupted_data = copy.deepcopy(infer.gene_dataset.X)
```
Comment (Edouard360) [inline scvi/metrics/imputation.py]: Not sure if that is memory efficient. (Before we did drop "on the fly", but there was no training of the model on the corrupted data, just forward pass on corrupted batches)

**Comment by Edouard360** [comment]
Yes exactly, the median over all cells and all genes. I'm not sure more helps, there is some variance across runs it seems, like I have had around 2.1 ~ 2.5 sometimes. In 250 epochs, the model might not yet be fully trained (ll decreases again I think), so maybe hte results stabilize more at the end of training

**Comment by Edouard360** [comment]
I will take a look at that + adapt notebooks according to the new slight changes in code

**Comment by Edouard360** [comment]
10 runs on different corrupted instance on Cortex, where at the end of each experiment, we use different values of n_samples for the median imputation score.

| 1  | 8 | 32 | 128 |
| - | - |- |- |
| 2.20968175 | 2.20870447 |2.20864177 | 2.20848131|

These are the values averaged over the 10 runs. The differences are of the order 0.001.

**Comment by Edouard360** [comment]
So we are actually better then ZINB-WAVE with enough epochs for training !

File: scvi/inference/variational_inference.py (hunk 1 above)
Comment (jeff-regier) [reply scvi/inference/variational_inference.py]: Yes, I think that is good since that's what in the paper, even though it's kind of an arbitrary metric.

File: scvi/metrics/imputation.py (hunk 4 above)
Comment (jeff-regier) [reply scvi/metrics/imputation.py]: Won't that double the memory use of the program? Does it work to "corrupt" the training data on the fly, in the training loop?

File: scvi/metrics/imputation.py (hunk 3 above)
Comment (jeff-regier) [reply scvi/metrics/imputation.py]: Are you trying to implement wake-sleep here? Or is this data augmentation?

**Comment by jeff-regier** [comment]
If we're only using hidden one layer for Cortex, it's pretty hard to improve much over zinb-wave. (Both models are pretty inflexible then.). A larger dataset might be better for comparison.

File: scvi/metrics/imputation.py (hunk 4 above)
Comment (Edouard360) [reply scvi/metrics/imputation.py]: Yes I agree, it is suboptimal in terms of space but not in term of clarity I think.

I don't see a very clean way to do it on the fly. We have to generate anyway 10% of new data. We could maybe somehow get the indices of the shuffled dataloaders in the training loop, but it would be computationally inefficient to overwrite the input for the corresponding indices at each loop iteration.

For a large dataset like brain large, romain only took a subsample of the dataset to run the imputation benchmark.

File: scvi/metrics/imputation.py (hunk 3 above)
Comment (Edouard360) [reply scvi/metrics/imputation.py]: I want all the dataloaders to have their corrupted equivalent.
For VAE, one data_loader is enough for training, but we still have labelled and unlabelled dataloaders for VAEC, and SVAEC
